components/heap/heap_trace_standalone.c

```diff
@@ -239,47 +291,32 @@ static void heap_trace_dump_base(bool internal_ram, bool psram)
         printf("(NB: Internal Buffer has overflowed, so trace data is incomplete.)\n");
     }
     printf("================================\n");
+
+    portEXIT_CRITICAL(&trace_mux);
 }
 
 /* Add a new allocation to the heap trace records */
-static IRAM_ATTR void record_allocation(const heap_trace_record_t *record)
+static IRAM_ATTR void record_allocation(const heap_trace_record_t *rAllocation)
 {
-    if (!tracing || record->address == NULL) {
+    if (!tracing || rAllocation->address == NULL) {
         return;
     }
 
     portENTER_CRITICAL(&trace_mux);
 
     if (tracing) {
 
+        // If buffer is full, pop off the oldest 
+        // record to make more space
         if (records.count == records.capacity) {
 
             records.has_overflowed = true;
 
-            /* Move the whole buffer back one slot.
-
-            This is a bit slow, compared to treating this buffer as a
-            ringbuffer and rotating a head pointer.
-
-            However, ringbuffer code gets tricky when we remove elements
-            in mid-buffer (for leak trace mode) while trying to keep
-            track of an item count that may overflow.
-            */
-            memmove(&records.buffer[0], &records.buffer[1],
-                sizeof(heap_trace_record_t) * (records.capacity -1));
-
-            records.count--;
+            linked_list_remove(&records, records.first);
         }
```

Comment:
This check is done again in `linked_list_add`. Maybe you could move this block at the beginning of `linked_list_add` instead of only adding the element if capacity is not reached. In that way `record_allocation` gets more readable and `linked_list_add` never fails and can therefore return void which makes sense since the return value is not checked line 319.